File: src/nercone_fonts/package.py
import os
import shutil
import tarfile
import zipfile
import subprocess
import tempfile
from typing import Optional, List, Dict

from .models import Format, License, Family
from .constants import Paths

class Archives:
    zip = "zip"
    sevenzip = "7z"
    gzip = "tar.gz"
    xz = "tar.xz"

    all = [zip, sevenzip, gzip, xz]

    tar = {gzip: "w:gz", xz: "w:xz"}

class Packager:
    def __init__(self, name: str, families: List[Family], license: License, source: str = Paths.files, directory: str = Paths.dist, logger=None):
        self.name = name
        self.families = list(families)
        self.license = license
        self.source = source
        self.directory = directory
        self.logger = logger

    def note(self, message: str):
        if self.logger:
            self.logger.info(message)

    def contents(self) -> Dict[str, str]:
        entries = {self.license.filename: self.license.filepath}

        for family in self.families:
            for style in family.styles():
                for format in Format:
                    filename = "{}-{}.{}".format(family.filename, style.name(), format.extension())
                    folder = "{}/{}".format(format.group(), format.directory())
                    if not style.variable():
                        folder += "/Static"
                    entries["{}/{}".format(folder, filename)] = os.path.join(self.source, filename)

        return entries

    def missing(self) -> List[str]:
        return [path for path in self.contents().values() if not os.path.exists(path)]

    def package(self, formats: Optional[List[str]] = None) -> List[str]:
        os.makedirs(self.directory, exist_ok=True)

        contents = self.contents()
        absent = self.missing()
        if absent:
            raise FileNotFoundError("{} is not built yet: {} missing".format(self.name, len(absent)))

        written = []
        for format in (formats or Archives.all):
            path = os.path.join(self.directory, "{}.{}".format(self.name, format))
            if format == Archives.zip:
                self.compress(path, contents)
            elif format == Archives.sevenzip:
                self.collect(path, contents)
            else:
                self.archive(path, contents, Archives.tar[format])
            written.append(path)
            self.note("packaged {}".format(path))

        return written

    def compress(self, path: str, contents: Dict[str, str]):
        with zipfile.ZipFile(path, "w", zipfile.ZIP_DEFLATED, compresslevel=9) as archive:
            for name, source in sorted(contents.items()):
                archive.write(source, "{}/{}".format(self.name, name))

    def archive(self, path: str, contents: Dict[str, str], mode: str):
        with tarfile.open(path, mode) as archive:
            for name, source in sorted(contents.items()):
                archive.add(source, "{}/{}".format(self.name, name))

    def collect(self, path: str, contents: Dict[str, str]):
        if os.path.exists(path):
            os.remove(path)

        try:
            import py7zr
        except ImportError:
            py7zr = None

        if py7zr is not None:
            with py7zr.SevenZipFile(path, "w") as archive:
                for name, source in sorted(contents.items()):
                    archive.write(source, "{}/{}".format(self.name, name))
            return

        binary = shutil.which("7z") or shutil.which("7zz") or shutil.which("7za")
        if binary is None:
            raise RuntimeError("7z archives need either the py7zr package or the 7z command")

        with tempfile.TemporaryDirectory() as staging:
            for name, source in contents.items():
                target = os.path.join(staging, self.name, name)
                os.makedirs(os.path.dirname(target), exist_ok=True)
                shutil.copyfile(source, target)
            subprocess.run([binary, "a", "-mx=9", "-bso0", "-bsp0", os.path.abspath(path), self.name], cwd=staging, check=True)

File: src/package.rs
use std::collections::BTreeMap;
use std::fs::File;
use std::path::Path;

use crate::models::{Format, License, Family};
use crate::constants::Paths;

pub struct Archives;

#[allow(non_upper_case_globals)]
impl Archives {
    pub const xz: &'static str = "tar.xz";
}

pub struct Packager {
    pub name: String,
    pub families: Vec<Family>,
    pub license: License,
    pub source: String,
    pub directory: String,
}

impl Packager {
    pub fn new(name: String, families: Vec<Family>, license: License) -> Packager {
        Packager {
            name,
            families,
            license,
            source: Paths::files.to_string(),
            directory: Paths::dist.to_string(),
        }
    }

    pub fn family(family: Family) -> Packager {
        let (name, license) = (family.filename.clone(), family.license);
        Packager::new(name, vec![family], license)
    }

    pub fn collection(name: String, families: Vec<Family>) -> Packager {
        let license = families[0].license;
        Packager::new(name, families, license)
    }

    pub fn note(&self, message: &str) {
        println!("{}", message);
    }

    pub fn contents(&self) -> BTreeMap<String, String> {
        let mut entries = BTreeMap::new();
        entries.insert(self.license.filename.to_string(), self.license.filepath.to_string());

        for family in &self.families {
            for style in family.styles() {
                for format in Format::all() {
                    let filename = format!("{}-{}.{}", family.filename, style.name(), format.extension());
                    let mut folder = format!("{}/{}", format.group(), format.directory());
                    if !style.variable() {
                        folder += "/Static";
                    }
                    entries.insert(format!("{}/{}", folder, filename),
                                   Path::new(&self.source).join(&filename).to_string_lossy().into_owned());
                }
            }
        }

        entries
    }

    pub fn missing(&self) -> Vec<String> {
        self.contents().values()
            .filter(|path| !Path::new(path).exists())
            .cloned()
            .collect()
    }

    pub fn package(&self) -> String {
        std::fs::create_dir_all(&self.directory)
            .unwrap_or_else(|error| panic!("failed to create {}: {}", self.directory, error));

        let contents = self.contents();
        let absent = self.missing();
        if !absent.is_empty() {
            panic!("{} is not built yet: {} missing", self.name, absent.len());
        }

        let path = Path::new(&self.directory)
            .join(format!("{}.{}", self.name, Archives::xz))
            .to_string_lossy()
            .into_owned();
        self.archive(&path, &contents);
        self.note(&format!("packaged {}", path));
        path
    }

    pub fn archive(&self, path: &str, contents: &BTreeMap<String, String>) {
        let file = File::create(path)
            .unwrap_or_else(|error| panic!("failed to create {}: {}", path, error));
        let encoder = liblzma::write::XzEncoder::new(file, 9);
        let mut builder = tar::Builder::new(encoder);
        for (name, source) in contents {
            builder.append_path_with_name(source, format!("{}/{}", self.name, name))
                .unwrap_or_else(|error| panic!("failed to add {} to {}: {}", name, path, error));
        }
        builder.into_inner()
            .and_then(|encoder| encoder.finish())
            .unwrap_or_else(|error| panic!("failed to write {}: {}", path, error));
    }
}
